examples/shell/cmd_otcli.cpp

```diff
@@ -0,0 +1,168 @@
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <core/CHIPCore.h>
+#include <platform/CHIPDeviceLayer.h>
+
+#if CHIP_ENABLE_OPENTHREAD
+
+#include <stdio.h>
+
+#include <platform/ThreadStackManager.h>
+#include <shell/shell.h>
+#include <support/CHIPArgParser.hpp>
+#include <support/CHIPMem.h>
+#include <support/CodeUtils.h>
+
+#if CHIP_TARGET_STYLE_EMBEDDED
+#include <openthread/cli.h>
+#include <openthread/instance.h>
+#include <openthread/ip6.h>
+#include <openthread/link.h>
+#include <openthread/thread.h>
+#else
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#endif
+
+using namespace chip;
+using namespace chip::Shell;
+using namespace chip::Platform;
+using namespace chip::DeviceLayer;
+using namespace chip::Logging;
+using namespace chip::ArgParser;
+
+static chip::Shell::Shell sShellOtcliSubcommands;
+
+int cmd_otcli_help_iterator(shell_command_t * command, void * arg)
+{
+    streamer_printf(streamer_get(), "  %-15s %s\n\r", command->cmd_name, command->cmd_help);
+    return 0;
+}
+
+int cmd_otcli_help(int argc, char ** argv)
+{
+    sShellOtcliSubcommands.ForEachCommand(cmd_otcli_help_iterator, NULL);
+    return 0;
+}
+
+#if CHIP_TARGET_STYLE_EMBEDDED
+
+int cmd_otcli_dispatch(int argc, char ** argv)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+// From OT CLI internal lib, kMaxLineLength = 128
+#define kMaxLineLength 128
+    char buff[kMaxLineLength] = { 0 };
+    char * buff_ptr           = buff;
+    int i                     = 0;
+
+    VerifyOrExit(argc > 0, error = CHIP_ERROR_INVALID_ARGUMENT);
+
+    for (i = 0; i < argc; i++)
+    {
+        size_t arg_len = strlen(argv[i]);
+
+        /* Make sure that the next argument won't overflow the buffer */
+        VerifyOrExit(buff_ptr + arg_len < buff + kMaxLineLength, error = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+        strncpy(buff_ptr, argv[i], arg_len);
+        buff_ptr += arg_len;
+
+        /* Make sure that there is enough buffer for a space char */
+        if (buff_ptr + sizeof(char) < buff + kMaxLineLength)
+        {
+            strncpy(buff_ptr, " ", sizeof(char));
+            buff_ptr++;
+        }
+    }
+    buff_ptr = 0;
+
+    otCliConsoleInputLine(buff, buff_ptr - buff);
+exit:
+    return error;
+}
+
+#elif CHIP_TARGET_STYLE_UNIX
+
+int cmd_otcli_dispatch(int argc, char ** argv)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    int pid, status;
+    uid_t euid         = geteuid();
+    char ctl_command[] = "/usr/local/sbin/ot-ctl";
```

Comment:
This will probably needs more configurability in the future. Could use execvp().

Reply:
I switched to `execvp()`, but when running sudo, it still has trouble finding simply `ot-cli`, even if I add `export PATH=$PATH:/usr/local/sbin` to `/root/.bashrc`.

Reply:
That's a sudo policy biting:

```
sudo grep secure_path /etc/sudoers
Defaults secure_path=/usr/sbin:/usr/bin:/sbin:/bin
```

Reply:
Maybe stick a TODO for now to add some way to override (could just be a command line argument).